### PR TITLE
Add quota setting for correlations

### LINKED ISSUE #63955: Glue: Limit number of correlations that can be created
### Background

As a part of feature roll-out we should ensure that we have input bounds in place. At the moment there's no limit to number of created correlations. 

### Requirements ###

High level: _Define limits for correlations stored in the database_

#### Technical details

* Limits should be part of the configuration with no limit as default.
* The limit defines how many correlations can be defined per data source. Throw an exception when more is being added.

### PR DESCRIPTION
**Pull request by gelicia**
**What is this feature?**

This adds a global quota for correlations. 

**Why do we need this feature?**

Some users may want to limit the number of correlations that can be created. This gives them the config option to do so.

**Which issue(s) does this PR fix?**:

Fixes #63955

**Special notes for your reviewer:**

Please add the following to custom.ini to test this quota. Note: if the limit is lower than the number of provisioned correlations that are created, it will fail.

```
[quota]
enabled = true
global_correlations = -1
```

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.
- [ ] There are no known compatibility issues with older supported versions of Grafana, or plugins.
- [ ] It passes the [Hosted Grafana feature readiness review](https://docs.google.com/document/d/1QL9Ly8KnXzpb6ISbg49pTODRO5mhA5tkkfIZVX6pqQU/edit) for observability, scalability, performance, and security.

### CODE REVIEW

**Review by ifrost — APPROVED** [review]
Splendid ❤️‍🔥 

The description of the tasks mentions the limit should be per datasource, but I actually like global limit - simpler and more consistent. 

Tested with provisioning and adding from the UI / API. Works great! 🎉

**Comment by gelicia** [comment]
> The description of the tasks mentions the limit should be per datasource, but I actually like global limit - simpler and more consistent.

Ooops, that's right. I was spending so much time figuring out how quota should be implemented I missed this detail. Let me check with the correlations channel on if one single limit is okay

**Comment by ifrost** [comment]
> Ooops, that's right. I was spending so much time figuring out how quota should be implemented I missed this detail. Let me check with the correlations channel on if one single limit is okay

I think it's okay. When we defined requirements I think we didn't know that we already have usage quotas defined globally or per org. It seems better to have it the way you implemented it, it's more consistent and simpler.

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 1]
```diff
@@ -1358,6 +1358,10 @@ Sets a global limit on number of users that can be logged in at one time. Defaul
 
 Sets a global limit on number of alert rules that can be created. Default is -1 (unlimited).
 
+### global_correlations
+
+Sets a global limit on number of correlations that can be created. Default is -1 (unlimited).
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1363]: Typo: should be 'the number'. 

Also, 'correlations that can be created' is passive voice. Who is creating the correlations? I assume the user?

If it's the users, I'd say "Sets a global limit on the number of correlations you can create. The default value is `-1`, allowing unlimited correlations."

Comment (gelicia) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1363]: @chri2547 I followed what I thought was standard based on other quotas documented.

The user does create correlations. 

Screenshot 2023-03-22 at 11 58 29 AM
Screenshot 2023-03-22 at 11 59 26 AM

I can fix my quota, but I'm thinking reworking this file to fit the standards you're applying here will be a better path forward.

Comment (chri2547) [reply docs/sources/setup-grafana/configure-grafana/_index.md:1363]: @gelicia , You are correct that the file should be updated accordingly. I did introduce inconsistency, sorry about that. I think the words you used are fine as is, and if there's time we can work to update all the text.

**Pull request merged**